notifications_utils/clients/otel/otel_client.py

```diff
@@ -0,0 +1,179 @@
+import os
+
+from flask import Flask
+from opentelemetry import metrics
+from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
+from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
+from opentelemetry.processor.baggage import BaggageSpanProcessor
+from opentelemetry.sdk.metrics import MeterProvider
+from opentelemetry.sdk.metrics.export import (
+    ConsoleMetricExporter,
+    PeriodicExportingMetricReader,
+)
+from opentelemetry.sdk.resources import Resource
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter
+from opentelemetry.trace import get_tracer_provider, set_tracer_provider
+
+
+def init_otel_app(app: Flask) -> None:
+    """
+    Initialize the OpenTelemetry instrumentation for the Flask app.
+    """
+
+    if getattr(app, "_otel_instrumented", False):
+        app.logger.debug("OpenTelemetry instrumentation already applied, skipping.")
+        return
+
+    export_mode = app.config.get("OTEL_EXPORT_TYPE", "none").lower().strip()
+    metric_readers = []
+
+    if export_mode == "console":
+        app.logger.info("OpenTelemetry metrics and spans will be exported to console")
+        metric_readers.append(PeriodicExportingMetricReader(ConsoleMetricExporter()))
+        span_processor = BatchSpanProcessor(ConsoleSpanExporter())
+    elif export_mode == "otlp":
+        endpoint = app.config.get("OTEL_COLLECTOR_ENDPOINT", "localhost:4317")
+        app.logger.info("OpenTelemetry metrics and spans will be exported to OTLP collector at %s", endpoint)
+        otlp_exporter = OTLPMetricExporter(endpoint=endpoint, insecure=True)
+        metric_readers.append(PeriodicExportingMetricReader(otlp_exporter))
+
+        os.environ["OTEL_METRIC_EXPORT_INTERVAL"] = app.config.get("OTEL_METRIC_EXPORT_INTERVAL", "15s")
+        os.environ["OTEL_METRIC_EXPORT_TIMEOUT"] = app.config.get("OTEL_METRIC_EXPORT_TIMEOUT", "30s")
+
+        span_processor = BatchSpanProcessor(
+            OTLPSpanExporter(
+                endpoint=endpoint,
+                insecure=app.config.get("OTEL_COLLECTOR_INSECURE", True),
+            )
+        )
+    elif export_mode == "none":
+        app.logger.info("OpenTelemetry metrics and spans will not be exported")
+        return
+    else:
+        raise ValueError(f"Invalid OTEL_EXPORT_TYPE: {export_mode}. Expected 'console', 'otlp', or 'none'.")
+
+    resource = Resource.create(
+        {"service.name": os.getenv("NOTIFY_APP_NAME") or app.config.get("NOTIFY_APP_NAME") or "notifications"}
+    )
+
+    provider = MeterProvider(metric_readers=metric_readers, resource=resource)
+    metrics.set_meter_provider(provider)
+
+    set_tracer_provider(TracerProvider(resource=resource))
+
+    def regex_predicate(baggage_key: str) -> bool:
+        return baggage_key.startswith("^public-.+")
```

Comment:
This could do with some explanation. Why is it checking whether a string of some sort begins with a literal regex pattern? If it was cribbed from https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/processor/opentelemetry-processor-baggage#usage then I think that might be wrong - the code shows no special handling around regexes, it's just a pure string https://github.com/open-telemetry/opentelemetry-python-contrib/blob/80c357bb16b42c76763d85441e53dd4f6087d6c5/processor/opentelemetry-processor-baggage/src/opentelemetry/processor/baggage/processor.py#L61

Reply:
Previously I add all baggage getting added as an attribute. I thought that might be bad if we started passing PII in baggage so I added a regex. I didn't check to see if it works. I probably should have.

I could just remove the whole BaggageSpanProcessor if you want.

It's a case of me try to guess what might be useful to someone in the future rather than it being useful now.

Reply:
I've fixed the startswith. Let me know if you rather I removed entirely.

Reply:
I think I'd prefer `ALLOW_ALL_BAGGAGE_KEYS` - if we ever put PII in baggage then we've already screwed up and I would prefer to catch that during code review rather than relying on this regex to catch it